pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go

```diff
@@ -0,0 +1,506 @@
+						Message:            fmt.Sprintf("Error processing cloud request, err: %v", err),
+					},
+				},
+			}
+			// Always requeue the object if we end up here. We need to make sure
+			// we try to clean up the IP on the cloud
+			if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				cloudPrivateIPConfig, err = c.updateCloudPrivateIPConfigStatus(cloudPrivateIPConfig, status)
+				return err
+			}); err != nil {
+				return fmt.Errorf("Error updating CloudPrivateIPConfig: %s during delete operation, err: %v", name, err)
+			}
+			return fmt.Errorf("Error deleting IP address from node: %s for CloudPrivateIPConfig: %s, cloud err: %v", node.Name, name, cloudErr)
+		}
+		// Process real object deletion. We're using a finalizer, so it depends
+		// on this controller whether the object is finally deleted and removed
+		// from the store or not, hence don't check the store.
+		if !cloudPrivateIPConfig.ObjectMeta.DeletionTimestamp.IsZero() {
+			klog.Infof("CloudPrivateIPConfig: %s object has been marked for complete deletion", name)
+			if controllerutil.ContainsFinalizer(cloudPrivateIPConfig, cloudPrivateIPConfigFinalizer) {
+				// Everything has been cleaned up, remove the finalizer from the
+				// object and update so that the object gets removed. If it
+				// didn't get removed and we encountered an error we'll requeue
+				// it down below
+				controllerutil.RemoveFinalizer(cloudPrivateIPConfig, cloudPrivateIPConfigFinalizer)
+				klog.Infof("Cleaning up IP address and finalizer for CloudPrivateIPConfig: %s, deleting it completely", name)
+				return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+					_, err = c.updateCloudPrivateIPConfig(cloudPrivateIPConfig)
+					return err
+				})
+			}
+		}
+
+		klog.Infof("Deleted IP address from node: %s for CloudPrivateIPConfig: %s", node.Name, name)
+		if nodeToAdd != "" {
+			// Update the status here if we process an update so that it's
+			// evident to the consumer where we are in our sync and so that we
+			// can treat the remainder as an add in the next sync term, in case
+			// we fail from this moment on. If we only process a delete we'll
+			// update at the bottom.
+			status = &cloudnetworkv1.CloudPrivateIPConfigStatus{
+				Conditions: []metav1.Condition{
+					metav1.Condition{
+						Type:               string(cloudnetworkv1.Assigned),
+						Status:             metav1.ConditionUnknown,
+						ObservedGeneration: cloudPrivateIPConfig.Status.Conditions[0].ObservedGeneration + 1,
+						LastTransitionTime: metav1.Now(),
+						Reason:             cloudResponseReasonPending,
+					},
+				},
+			}
+			if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				cloudPrivateIPConfig, err = c.updateCloudPrivateIPConfigStatus(cloudPrivateIPConfig, status)
+				return err
+			}); err != nil {
+				return fmt.Errorf("Error updating CloudPrivateIPConfig: %s during delete operation, err: %v", name, err)
+			}
+		}
+	}
+
+	if nodeToAdd != "" {
```

Comment:
Multiple updates to the same IP. eg:

1. (IP1 is assigned to Node1)
2. Node1 goes down, SDN assigns IP1 to Node2
3. CNCC starts deleting IP1 from Node1...
4. Node2 goes down, SDN assigns IP1 to Node 3
5. CNCC finishes deleting IP1 from Node1
6. CNCC adds IP1 to Node2
7. CNCC deletes IP1 from Node2
8. CNCC adds IP1 to Node3

In theory, it should not do steps 6 and 7, but with the current code it will.

This may not be that big a deal.

Reply:
Not sure about that, we're using a queue informer and I thought that the delta FIFO will make sure the events are coalesced so that we only get the latest update, i.e: the following:

1. (IP1 is assigned to Node1)
2. Node1 goes down, SDN assigns IP1 to Node2
3. CNCC starts deleting IP1 from Node1...
4. Node2 goes down, SDN assigns IP1 to Node 3
5. CNCC finishes deleting IP1 from Node1
6. CNCC adds IP1 to Node3, because it gets the latest update on the object

Might be worth an experiment. I will see if I can formulate this as a unit test, otherwise I will manually test it out by adding a sleep in the sync.

Reply:
I've just tested this and what I stated seems to be correct: the updates are coalesced, so we will only get the latest update in case we are slow and happen to have multiple updates getting "stacked up".

Reply:
Never mind what I said, I was confused and wrote a dubious test which didn't verify anything of importance. I am not sure this might be very relevant in reality, but it's a real corner case to take into account. 

I need to think about it, but I suspect the solution is to XOR delete and add for update operations so that they won't be attempted during the same sync, that should have the example above solved by the update events being coalesced. I just need to think about other corner cases which might be impacted negatively by this, and make sure that is handled.